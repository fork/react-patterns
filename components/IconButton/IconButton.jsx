import React from 'react';
import PropTypes from 'prop-types';

import StyledIconButton from './IconButton.style';

import Icon from '../Icon';

const IconButton = ({ href, icon, size, disabled, className, label, ...props }) => (
  <StyledIconButton
    as={href ? 'a' : 'button'}
    type={href ? undefined : 'button'}
    className={className}
    disabled={disabled}
    aria-disabled={disabled}
    title={label}
    aria-label={label}
    size={size}
    {...props}
  >
    <Icon icon={icon} size={size} />
  </StyledIconButton>
);

IconButton.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node,
  disabled: PropTypes.bool,
  /**
   * Accessibility label
   */
  label: PropTypes.string.isRequired,
  variant: PropTypes.oneOf(['primary', 'secondary']),
  size: PropTypes.oneOf(['small', 'large']),
  icon: PropTypes.string.isRequired,
  /**
   * Defines the tag of the component. 'a' or 'button'.
   */
  href: PropTypes.string
};

IconButton.defaultProps = {
  className: null,
  children: null,
  disabled: false,
  size: 'small',
  variant: 'primary',
  href: null
};

export default IconButton;
