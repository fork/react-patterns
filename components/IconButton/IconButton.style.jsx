import styled, { css } from 'styled-components';

import { color, variant } from '../../stylesheets';

const StyledIconButton = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  outline: none;
  text-transform: uppercase;
  margin: 0;
  text-align: center;
  flex-shrink: 0;
  cursor: pointer;
  border-radius: 50%;

  &[disabled] {
    pointer-events: none;
    opacity: 0.5;
  }

  /* Variant Prop */
  ${variant({
    primary: css`
      color: ${color('white')};
      background: ${color('primary')};
    `,
    secondary: css`
      color: ${color('white')};
      background: ${color('secondary')};
    `
  })}

  /* Size Prop */
  ${variant({
    prop: 'size',
    small: css`
      width: 35px;
      height: 35px;
    `,
    large: css`
      width: 50px;
      height: 50px;
    `
  })}
`;

export default StyledIconButton;
