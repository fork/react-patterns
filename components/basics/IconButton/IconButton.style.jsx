import styled, { css } from 'styled-components';

import { color } from '../../../stylesheets';

const StyledIconButton = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  outline: none;
  text-transform: uppercase;
  margin: 0;
  text-align: center;
  flex-shrink: 0;
  cursor: pointer;
  border-radius: 50%;
  color: ${color('white')};
  background: ${color('primary')};
  width: 30px;
  height: 30px;

  &[disabled] {
    pointer-events: none;
    opacity: 0.5;
  }

  ${props =>
    props.size === 'large' &&
    css`
      width: 50px;
      height: 50px;
    `}

  ${props =>
    props.variant === 'secondary' &&
    css`
      color: ${color('white')};
      background: ${color('secondary')};
    `}
`;

export default StyledIconButton;
