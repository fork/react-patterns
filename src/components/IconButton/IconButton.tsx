import React from 'react';

import StyledIconButton from './IconButton.style';

import Icon, { IconTypes } from '../Icon';

export type IconButtonProps = {
  icon: IconTypes;
  size?: 'small' | 'large';
  variant?: 'primary' | 'secondary';
  label: string;
  className?: string;
  disabled?: boolean;
  onClick?: () => void;
};

const IconButton = ({
  icon,
  size = 'small',
  disabled,
  className,
  variant = 'primary',
  onClick,
  label
}: IconButtonProps) => (
  <StyledIconButton
    type="button"
    className={className}
    disabled={disabled}
    aria-disabled={disabled}
    title={label}
    aria-label={label}
    variant={variant}
    onClick={onClick}
    size={size}
  >
    <Icon icon={icon} size={size} />
  </StyledIconButton>
);

export default IconButton;
