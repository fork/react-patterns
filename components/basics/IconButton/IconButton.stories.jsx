import React from 'react';

import IconButton from './IconButton';

export default {
  title: 'Basics/IconButton',
  component: IconButton
};

export const Small = () => (
  <>
    <IconButton icon="date" label="This is a label" variant="primary" />

    <IconButton icon="date" label="This is a label" variant="secondary" />
  </>
);

export const Large = () => (
  <>
    <IconButton icon="date" size="large" label="This is a label" variant="primary" />
    <IconButton icon="date" size="large" label="This is a label" variant="secondary" />
  </>
);

export const asLink = () => (
  <>
    <IconButton
      icon="date"
      variant="primary"
      href="https://wwww.fork.de"
      label="This is a label"
      target="_blank"
    />
    <IconButton
      icon="date"
      variant="secondary"
      size="large"
      href="https://wwww.fork.de"
      label="This is a label"
      target="_blank"
    />
  </>
);
