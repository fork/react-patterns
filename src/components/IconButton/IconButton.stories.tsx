import React from 'react';

import IconButton from './IconButton';

export default {
  title: 'Components/IconButton',
  component: IconButton
};

export const Small: React.FC = () => (
  <>
    <IconButton icon="date" label="This is a label" variant="primary" />
    <IconButton icon="date" label="This is a label" variant="secondary" />
  </>
);

export const Large: React.FC = () => (
  <>
    <IconButton icon="date" size="large" label="This is a label" variant="primary" />
    <IconButton icon="date" size="large" label="This is a label" variant="secondary" />
  </>
);
