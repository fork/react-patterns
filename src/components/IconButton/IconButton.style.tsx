import styled, { css } from 'styled-components';

import { color, variant } from '../../stylesheets';
import { IconButtonProps } from './IconButton';

const StyledIconButton = styled.button<Pick<IconButtonProps, 'size' | 'variant'>>`
  display: flex;
  align-items: center;
  justify-content: center;
  outline: none;
  text-transform: uppercase;
  margin: 0;
  text-align: center;
  flex-shrink: 0;
  cursor: pointer;
  border-radius: 50%;

  &[disabled] {
    pointer-events: none;
    opacity: 0.5;
  }

  /* Variant Prop */
  ${variant({
    primary: css`
      color: ${color('white')};
      background: ${color('primary')};
    `,
    secondary: css`
      color: ${color('black')};
      background: ${color('secondary')};
    `
  })}

  /* Size Prop */
  ${variant({
    prop: 'size',
    small: css`
      width: 35px;
      height: 35px;
    `,
    large: css`
      width: 50px;
      height: 50px;
    `
  })}
`;

export default StyledIconButton;
