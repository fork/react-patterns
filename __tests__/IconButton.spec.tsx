import React from 'react';
import { shallow } from 'enzyme';

import IconButton from '../src/components/IconButton';

describe('IconButton', () => {
  test('IconButton should render correctly', () => {
    const component = shallow(<IconButton icon="close" label="Close" />);

    expect(component).toMatchSnapshot();
  });

  test('Secondary IconButton should render correctly', () => {
    const component = shallow(<IconButton icon="close" variant="secondary" label="Close" />);

    expect(component).toMatchSnapshot();
  });
});
