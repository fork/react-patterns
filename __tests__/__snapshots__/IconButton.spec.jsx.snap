// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton IconButton should render correctly 1`] = `
<styled.button
  aria-disabled={false}
  aria-label="Close"
  as="button"
  className={null}
  disabled={false}
  size="small"
  title="Close"
  type="button"
  variant="primary"
>
  <Icon
    icon="close"
    size="small"
  />
</styled.button>
`;

exports[`IconButton Secondary IconButton should render correctly 1`] = `
<styled.button
  aria-disabled={false}
  aria-label="Close"
  as="button"
  className={null}
  disabled={false}
  size="small"
  title="Close"
  type="button"
  variant="secondary"
>
  <Icon
    icon="close"
    size="small"
  />
</styled.button>
`;
